sway-lib-std/src/storage.sw

```diff
@@ -50,34 +51,33 @@ pub fn get<T>(key: b256) -> T {
         asm(l: loaded_word) { l: T }
     } else {
         // If reference type, then it can be more than a word. Loop over every 32
-        // bytes and read sequentially.
+        // bytes and read sequentially.  NOTE: we are leaking this value on the heap.
         let mut size_left = __size_of::<T>();
         let mut local_key = key;
 
-        // Keep track of the base pointer for the final result
-        let result_ptr = stack_ptr();
+        // Allocate a buffer for the result.  It needs to be a multiple of 32 bytes so we can make
+        // 'quad' storage reads without overflowing.
+        //
+        // XXX This fails for tests that exercise the loop below unless we allocate an extra byte!
+        // This must be a bug?!
+        //
+        let result_ptr = alloc(((size_left + 31) & 0xffffffe0) + 1);
```

Comment:
If we don't allocate that extra byte then this will fail with a `MemoryOverflow` from the VM.

Reply:
I could not figure it out. I played around with it a bit and here are some observations:
* Even the case with `b256` (so the loop is not really exercised) is failing, but not all reference types are failing!
* The issue is when _reading_ the final value, so all the steps work fine until we get to the `asm` block at the end.
* Reading from `result_ptr - 1` at the end instead of `result_ptr` does not overflow, but shows that the data is there correctly (except for the first byte which is always 0 because we're reading `result_ptr - 1` and not `result_ptr`. So the data is there but we can't read it for some reason? I also tried replacing the `asm` block with `mem::read` and got the same behavior (expectedly).

Reply:
I even logged the size allocated and it's correct (` = 32` in the case of `b256`).

Reply:
And for whatever reason, this works:
```rust
        let result_ptr = alloc(1);
        let result_ptr = realloc(result_ptr, 1, ((size_left + 31) & 0xffffffe0) + 0);
```

Reply:
There definitely is a minor weirdness with `ALOC` and `$hp` in that `$hp` always points to the last _unallocated_ byte, though `std::alloc::alloc()` accounts for that.  https://github.com/FuelLabs/fuel-vm/issues/137

The heap grows _down_, so the above snippet doesn't surprise me, it's pretty much the same as `size + 1`.

I was thinking it was only reads that used the loop, but it must be reads which are multiples of 32 bytes and need the entire buffer.  I think most of those tests are actually only 16 bytes.

Reading from `result_ptr - 1` should _not_ work, this is the biggest surprise to me.  It's almost as if the change in the issue I proposed above has been made.  :thinking:  If that were true `Vec` should be broken, as the primary user of `alloc()`.